staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -219,14 +220,23 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *admiss
 	if err != nil {
 		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
 	}
-	trace := utiltrace.New("Call mutating webhook",
-		utiltrace.Field{"configuration", configurationName},
-		utiltrace.Field{"webhook", h.Name},
-		utiltrace.Field{"resource", attr.GetResource()},
-		utiltrace.Field{"subresource", attr.GetSubresource()},
-		utiltrace.Field{"operation", attr.GetOperation()},
-		utiltrace.Field{"UID", uid})
-	defer trace.LogIfLong(500 * time.Millisecond)
+	traceStep, ok := genericapirequest.TraceFrom(ctx)
```

Comment:
a trace in the context is strange IMO. `LogIfLong` will print the total time and the dominating step when called. And it is called when this stack frame is closed. The creator of the trace (the one who put it into the context) would also call `LogIfLong`. But `LogIfLong` gives confusing output when called twice.

Reply:
oh, ic. You only defer the `LogIfLong` if it is not from the context.

Reply:
Still I am not convinced a trace in a context really makes sense, at least not without representing the hierarchy as a tree, e.g. with nested traces. The goal is to delay the total time output, isn't it? Some code further up the call chain will destroy proper trace output for inner trace steps because the total time gets a new sense now.

Reply:
@sttts These changes help to avoid creating new traces further down the call chain.

Reply:
What we really want is clever span handling, that's what nested traces are:
```
trace, ctx := utiltrace.NewNestedInContext(ctx, "request")
defer trace.LogIfLong(60 * time.Second)
func (ctx context.Context){
    trace, ctx := utiltrace.NewNestedInContext(ctx, "admission")
    defer trace.LogIfLong(2 * time.Second)
    func (ctx context.Context){
        trace, ctx := utiltrace.NewNestedInContext(ctx, "webhook")
        defer trace.LogIfLong(500 * time.Millisecond)
    }(ctx)
}(ctx)
```

If only the webhook is slow (slower than 500ms), I only want to see logging for the inner span.

If the admission is slow and the reason is the webhook taking more than 2s, I want to see both spans printed out:
```
Admission was slow (total time 2.8s): webhook was slow (total time 2.5s):
[admission [webhook [web-hook-1 0.5s] [web-hook-2 2s] 2.5s] 2.8s]
```